File: app/src/main/java/pl/trikimusic/controller/core/bluetooth/MultiDeviceArbitrationPolicy.kt
package pl.trikimusic.controller.core.bluetooth

import pl.trikimusic.controller.domain.model.MultiDeviceArbitrationMode

object MultiDeviceArbitrationPolicy {
    const val RECENT_PLAYBACK_WINDOW_MILLIS = 3 * 60 * 1000L
    const val IDLE_PLAYBACK_WINDOW_MILLIS = 10 * 60 * 1000L

    const val ACTIVE_PLAYBACK_DELAY_MILLIS = 0L
    const val FOREGROUND_OR_ACTIVE_USER_DELAY_MILLIS = 100L
    const val RECENT_PLAYBACK_DELAY_MILLIS = 200L
    const val STALE_PLAYBACK_DELAY_MILLIS = 500L
    const val IDLE_YIELD_DELAY_MILLIS = 1000L

    fun shouldAttemptConnection(
        mode: MultiDeviceArbitrationMode,
        isMediaPlaying: Boolean,
        isUserActiveOrForeground: Boolean,
    ): Boolean = when (mode) {
        MultiDeviceArbitrationMode.ONLY_WHEN_PLAYING -> isMediaPlaying || isUserActiveOrForeground
        MultiDeviceArbitrationMode.ALWAYS_CONNECT -> true
        MultiDeviceArbitrationMode.MEDIA_PRIORITY -> true
    }

    fun calculateConnectionDelay(
        mode: MultiDeviceArbitrationMode,
        isMediaPlaying: Boolean,
        lastPlaybackTimeMillis: Long?,
        isUserActiveOrForeground: Boolean,
        nowMillis: Long,
    ): Long {
        if (mode == MultiDeviceArbitrationMode.ALWAYS_CONNECT) return 0L
        if (mode == MultiDeviceArbitrationMode.ONLY_WHEN_PLAYING) return 0L

        // MediaPriority mode:
        if (isMediaPlaying) return ACTIVE_PLAYBACK_DELAY_MILLIS
        if (isUserActiveOrForeground) return FOREGROUND_OR_ACTIVE_USER_DELAY_MILLIS

        if (lastPlaybackTimeMillis != null) {
            val elapsed = (nowMillis - lastPlaybackTimeMillis).coerceAtLeast(0L)
            if (elapsed <= RECENT_PLAYBACK_WINDOW_MILLIS) return RECENT_PLAYBACK_DELAY_MILLIS
            if (elapsed <= IDLE_PLAYBACK_WINDOW_MILLIS) return STALE_PLAYBACK_DELAY_MILLIS
        }

        return IDLE_YIELD_DELAY_MILLIS
    }

    fun shouldYieldConnection(
        mode: MultiDeviceArbitrationMode,
        isMediaPlaying: Boolean,
        connectedDurationWithoutMediaMillis: Long,
    ): Boolean {
        if (isMediaPlaying) return false
        if (mode == MultiDeviceArbitrationMode.MEDIA_PRIORITY || mode == MultiDeviceArbitrationMode.ONLY_WHEN_PLAYING) {
            return connectedDurationWithoutMediaMillis >= 10_000L
        }
        return false
    }
}

File: windows/TrikiMusicController.Windows/Core/MultiDeviceArbitrationPolicy.cs
using TrikiMusicController_Windows.Models;

namespace TrikiMusicController_Windows.Core;

public static class MultiDeviceArbitrationPolicy
{
    public static readonly TimeSpan RecentPlaybackWindow = TimeSpan.FromMinutes(3);
    public static readonly TimeSpan IdlePlaybackWindow = TimeSpan.FromMinutes(10);

    public static readonly TimeSpan ActivePlaybackDelay = TimeSpan.Zero;
    public static readonly TimeSpan ForegroundOrActiveUserDelay = TimeSpan.FromMilliseconds(100);
    public static readonly TimeSpan RecentPlaybackDelay = TimeSpan.FromMilliseconds(200);
    public static readonly TimeSpan StalePlaybackDelay = TimeSpan.FromMilliseconds(500);
    public static readonly TimeSpan IdleYieldDelay = TimeSpan.FromMilliseconds(1000);

    public static bool ShouldAttemptConnection(
        MultiDeviceArbitrationMode mode,
        bool isMediaPlaying,
        bool isUserActiveOrForeground) =>
        mode switch
        {
            MultiDeviceArbitrationMode.OnlyWhenPlaying => isMediaPlaying || isUserActiveOrForeground,
            MultiDeviceArbitrationMode.AlwaysConnect => true,
            MultiDeviceArbitrationMode.MediaPriority => true,
            _ => true,
        };

    public static TimeSpan CalculateConnectionDelay(
        MultiDeviceArbitrationMode mode,
        bool isMediaPlaying,
        DateTimeOffset? lastPlaybackTime,
        bool isUserActiveOrForeground,
        DateTimeOffset now)
    {
        if (mode == MultiDeviceArbitrationMode.AlwaysConnect) return TimeSpan.Zero;
        if (mode == MultiDeviceArbitrationMode.OnlyWhenPlaying) return TimeSpan.Zero;

        // MediaPriority mode:
        if (isMediaPlaying) return ActivePlaybackDelay;
        if (isUserActiveOrForeground) return ForegroundOrActiveUserDelay;

        if (lastPlaybackTime is DateTimeOffset lastTime)
        {
            var elapsed = now - lastTime;
            if (elapsed < TimeSpan.Zero) elapsed = TimeSpan.Zero;

            if (elapsed <= RecentPlaybackWindow) return RecentPlaybackDelay;
            if (elapsed <= IdlePlaybackWindow) return StalePlaybackDelay;
        }

        return IdleYieldDelay;
    }

    public static bool ShouldYieldConnection(
        MultiDeviceArbitrationMode mode,
        bool isMediaPlaying,
        TimeSpan connectedDurationWithoutMedia)
    {
        if (isMediaPlaying) return false;
        if (mode is MultiDeviceArbitrationMode.MediaPriority or MultiDeviceArbitrationMode.OnlyWhenPlaying)
        {
            return connectedDurationWithoutMedia >= TimeSpan.FromSeconds(10);
        }
        return false;
    }
}
